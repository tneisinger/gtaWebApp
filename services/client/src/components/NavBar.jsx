import React from 'react';
import { Navbar, Nav, NavItem } from 'react-bootstrap';
import { LinkContainer } from 'react-router-bootstrap';

const NavBar = (props) => (
  <Navbar inverse collapseOnSelect>
    <Navbar.Header>
      <Navbar.Brand>
        <LinkContainer to="/">
          <span>{props.title}</span>
        </LinkContainer>
      </Navbar.Brand>
      <Navbar.Toggle />
    </Navbar.Header>
    <Navbar.Collapse>
      <Nav>
        <LinkContainer to="/calendar">
          <NavItem eventKey={1}>Calendar</NavItem>
        </LinkContainer>
        <LinkContainer to="/budget">
          <NavItem eventKey={2}>Budget</NavItem>
        </LinkContainer>
        <LinkContainer to="/expenses">
          <NavItem eventKey={3}>Expenses</NavItem>
        </LinkContainer>
      </Nav>
      <Nav pullRight>
        <LinkContainer to="/register">
          <NavItem eventKey={1}>Register</NavItem>
        </LinkContainer>
        <LinkContainer to="/login">
          <NavItem eventKey={2}>Log In</NavItem>
        </LinkContainer>
        <LinkContainer to="/logout">
          <NavItem eventKey={3}>Log Out</NavItem>
        </LinkContainer>
      </Nav>
    </Navbar.Collapse>
  </Navbar>
)

export default NavBar;
