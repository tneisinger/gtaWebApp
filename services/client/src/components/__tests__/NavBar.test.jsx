import React from 'react';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { MemoryRouter as Router } from 'react-router-dom';

import NavBar from '../NavBar';

describe('The NavBar component', () => {

  let props = {
    onAuthBtnClick: undefined,
    userIsAdmin: undefined,
  };

  // Initialize a variable that will hold the mounted NavBar component
  let mountedNavBar;

  // Initialize a variable that will hold the navbar tree for snapshots
  let navbarTree;

  // Define a function that will always return the mounted NavBar component.
  // If the mountedNavBar hasn't been created yet for the current test,
  // create it and then return it.
  const navbar = () => {
    if (!mountedNavBar) {
      mountedNavBar = mount(
        <Router location='/'><NavBar {...props} /></Router>
      );
    }
    return mountedNavBar;
  }

  const tree = () => {
    if (!navbarTree) {
      navbarTree = renderer.create(
        <Router location='/'><NavBar {...props} /></Router>
      ).toJSON()
    }
    return navbarTree;
  }

  beforeEach(() => {
    // Reset the prop values for each test
    props = {
      onAuthBtnClick: undefined,
      userIsAdmin: undefined,
    };

    // throw away the old mountedNavBar
    mountedNavBar = undefined;

    // throw away the old navbarTree
    navbarTree = undefined;
  });

  describe('when userIsAdmin is set to true', () => {
    beforeEach(() => {
      props.userIsAdmin = true;
    });

    it('should render the calendar, budget, and expenses links', () => {
      const navbarNavs = navbar().find('.navbar-nav');
      expect(navbarNavs.length).toBeGreaterThan(1);
      expect(navbarNavs.first().find('li').length).toBeGreaterThan(1);
    });

    it('the text of the login/logout button should be "Logout"', () => {
      const authBtn = navbar().find('button.auth-btn');
      expect(authBtn.text()).toBe('Logout');
    });

    it('should render properly', () => {
      expect(tree()).toMatchSnapshot();
    });

  });

  describe('when userIsAdmin is set to false', () => {
    beforeEach(() => {
      props.userIsAdmin = false;
    });

    it('should NOT render the calendar, budget, and expenses links', () => {
      const navbarNavs = navbar().find('.navbar-nav');
      expect(navbarNavs.length).toBe(1);
      expect(navbarNavs.find('button').length).toBe(1);
    });

    it('the text of the login/logout button should be "Login"', () => {
      const authBtn = navbar().find('button.auth-btn');
      expect(authBtn.text()).toBe('Login');
    });

    it('should render properly', () => {
      expect(tree()).toMatchSnapshot();
    });
  });

  describe('when userIsAdmin is set to null', () => {
    beforeEach(() => {
      props.userIsAdmin = null;
    });

    it('should NOT render the admin links or the auth button', () => {
      const navbarNavs = navbar().find('.navbar-nav');
      expect(navbarNavs.length).toBe(0);
      expect(navbarNavs.find('button').length).toBe(0);
    });

    it('the login/logout button should not be rendered', () => {
      const authBtn = navbar().find('button.auth-btn');
      expect(authBtn.length).toBe(0);
    });

    it('should render properly', () => {
      expect(tree()).toMatchSnapshot();
    });
  });

});
