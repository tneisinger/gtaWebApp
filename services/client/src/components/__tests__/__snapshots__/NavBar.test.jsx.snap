// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar renders a snapshot properly 1`] = `
<nav
  className="navbar navbar-inverse"
>
  <div
    className="container"
  >
    <div
      className="navbar-header"
    >
      <span
        className="navbar-brand  active"
        href="/"
        onClick={[Function]}
        style={Object {}}
      >
        Hello, World!
      </span>
      <button
        className="navbar-toggle collapsed"
        onClick={[Function]}
        type="button"
      >
        <span
          className="sr-only"
        >
          Toggle navigation
        </span>
        <span
          className="icon-bar"
        />
        <span
          className="icon-bar"
        />
        <span
          className="icon-bar"
        />
      </button>
    </div>
    <div
      aria-expanded={null}
      className="navbar-collapse collapse"
    >
      <ul
        className="nav navbar-nav"
        role={null}
      >
        <li
          className=""
          role="presentation"
          style={undefined}
        >
          <a
            href="/calendar"
            onClick={[Function]}
            onKeyDown={[Function]}
          >
            Calendar
          </a>
        </li>
        <li
          className=""
          role="presentation"
          style={undefined}
        >
          <a
            href="/budget"
            onClick={[Function]}
            onKeyDown={[Function]}
          >
            Budget
          </a>
        </li>
        <li
          className=""
          role="presentation"
          style={undefined}
        >
          <a
            href="/expenses"
            onClick={[Function]}
            onKeyDown={[Function]}
          >
            Expenses
          </a>
        </li>
      </ul>
      <ul
        className="nav navbar-nav navbar-right"
        role={null}
      >
        <li
          className=""
          role="presentation"
          style={undefined}
        >
          <a
            href="/register"
            onClick={[Function]}
            onKeyDown={[Function]}
          >
            Register
          </a>
        </li>
        <li
          className=""
          role="presentation"
          style={undefined}
        >
          <a
            href="/login"
            onClick={[Function]}
            onKeyDown={[Function]}
          >
            Log In
          </a>
        </li>
        <li
          className=""
          role="presentation"
          style={undefined}
        >
          <a
            href="/logout"
            onClick={[Function]}
            onKeyDown={[Function]}
          >
            Log Out
          </a>
        </li>
      </ul>
    </div>
  </div>
</nav>
`;
